### PR TITLE
Disable missing CRDs in 2.x

### LINKED ISSUE #1448: Add back dynamically configured controllers
In previous iterations of KIC 2.0 we had functionality that dynamically enabled/disabled controllers based on the presence of relevant CRDs. This got lost in some refactoring of the controller manager, and the purpose of this issue is to re-add this functionality.

Acceptance Criteria:
- [x] controllers for 3rd party APIs (knative, apiext, e.t.c.) should be configured for dynamic enablement based on the presence of relevant CRDs
- [x] controllers which are automatically enabled/disabled should have their status logged in the controller manager log output.

### PR DESCRIPTION
**Pull request by rainest**
**What this PR does / why we need it**:

- Refactors the CRD detection utility to work with any CRD.
- Disables the KongClusterPlugin controller if it is not present. We maybe need this less in 2.x given the explicit flag and that you have to intentionally remove the CRD when installing KIC, but it probably should be in still to smooth 1.x -> 2.x upgrades.
- Updates defaults for Knative controller. Since it is automatically disabled if not present, we can enable it by default.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #1448 

**Special notes for your reviewer**:

#1396 previously added a utility for this specific to Knative. This PR makes the utility generic and uses it for KongClusterPlugins also. This utility does differ from the 1.x utility (see [the CRD detection function](https://github.com/Kong/kubernetes-ingress-controller/blob/1.3.1/pkg/util/ingressapi.go#L33-L46) and [example usage](https://github.com/Kong/kubernetes-ingress-controller/blob/1.3.1/cli/ingress-controller/main.go#L373-L387)), though I don't think there's any reason we need to make the 2.x version match it--is anyone aware of a meaningful difference between the two strategies?

#1448 suggests there may be other CRDs we want to catch with this. Do we? There wasn't anything else like this in 1.x AFAIK. We may want to disable the apiextensions and/or networking v1beta1 Ingress controllers in the future for K8S versions that lack them, but I'm not sure we've committed to that yet.

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [ ] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

File: railgun/manager/run.go [hunk 1]
```diff
@@ -283,9 +274,36 @@ func Run(ctx context.Context, c *config.Config) error {
...
+				Log:              ctrl.Log.WithName("controllers").WithName("KongClusterPlugin"),
+				Scheme:           mgr.GetScheme(),
+				Proxy:            prx,
+				IngressClassName: c.IngressClassName,
+			},
+		}
+		controllers = append(controllers, controller)
+	} else {
+		setupLog.Info(`kongclusterplugins.configuration.konghq.com v1beta1 CRD not available on cluster.
+		Disabling KongClusterPlugin controller`)
```
Comment (ccfishk) [inline railgun/manager/run.go:297]: If I understand correctly, in 2.0, KongClusterPlugin is disabled when CRD does not exist at first place.

1.3.X mechanism varies, clusterplugin CR sync block until CRD get deployed, and the controller will be enabled after that.

File: railgun/manager/run.go [hunk 2]
```diff
@@ -297,6 +315,9 @@ func Run(ctx context.Context, c *config.Config) error {
 			},
 		}
 		controllers = append(controllers, controller)
+	} else {
+		setupLog.Info(`ingresses.networking.internal.knative.dev v1alpha1 CRD not available on cluster.
+		Disabling Knative controller`)
```
Comment (ccfishk) [inline railgun/manager/run.go:320]: configure Knative Controller as enabled
disable it if CRD does not exist
re-enable controller when CRD get deployed. 
I think this is consistent with 1.3.0 behavior.

File: railgun/internal/ctrlutils/utils.go [hunk 3]
```diff
@@ -133,14 +133,9 @@ func IsIngressClassSpecConfigured(obj client.Object, expectedIngressClassName st
...
-func KnativeCRDExist(client client.Client) bool {
-	knativeGVR := schema.GroupVersionResource{
-		Group:    knative.SchemeGroupVersion.Group,
-		Version:  knative.SchemeGroupVersion.Version,
-		Resource: "ingresses",
-	}
-	_, err := client.RESTMapper().KindFor(knativeGVR)
+// CRDExists returns false if CRD does not exist
+func CRDExists(client client.Client, gvr schema.GroupVersionResource) bool {
+	_, err := client.RESTMapper().KindFor(gvr)
```
Comment (ccfishk) [inline railgun/internal/ctrlutils/utils.go:138]: super ! makes this function more generic.

**Review by ccfishk — APPROVED** [review]
LGTM. 

Thanks for taking care of this

**Review by shaneutt — APPROVED** [review]
LGTM once Jimin's comments are addressed and CI is passing :+1:

**Review by shaneutt — COMMENTED** [review]
@rainest non-blocker, but how would you feel about managing core ingresses with this update as well?

We previously had code that would disable netv1, or apiextv1beta1 automatically in order to better support some older versions of Kubernetes (which wouldn't have a copy of netv1.Ingress by default).

If we don't do this, I don't think https://github.com/Kong/kubernetes-ingress-controller/issues/1449 is fully resolved because there will be version situations (I think k8s 1.16 is one of them) where our controller will fail to deploy because it can't find v1.Ingress and there's no flag in place to fix this?

If you don't have time for it, I can take this on.

**Comment by rainest** [comment]
> @rainest non-blocker, but how would you feel about managing core ingresses with this update as well?

I think this is quite doable--I thought the Ingress version negotiation in 1.x was something different for some reason. I want to double-check what we did there for the different versions to be sure, but assuming it's not or it's close enough that it can be easily adapted, I'll add it to this.

**Comment by rainest** [comment]
b43acadc adds something similar to https://github.com/Kong/kubernetes-ingress-controller/blob/1.3.1/pkg/util/ingressapi.go for 2.x. The 2.x variant differs because we no longer use discovery clients. It's also explicitly for Ingress APIs only, whereas the original `NegotiateResourceAPI()` suggested generic API support, but was de facto Ingress-only because it used the `IngressAPI` type in its arguments and return values.

We will presumably need a generic API version negotiator in the future as we add new versions of our CRDs, which I expect will happen at some point. I didn't design for that in the proposed 2.x implementation because we don't yet have a use for anything other than Ingress and designing generic functions before you can actually use them for more than one type is hard.

I'm not immediately sure how to add unit tests for the new system for lack of client-go familiarity. Would anyone with more knowledge be able to assist me in setting up a mock environment similar to what we have in https://github.com/Kong/kubernetes-ingress-controller/blob/1.3.1/pkg/util/ingressapi_test.go but suitable for use with `client.RESTMapper()`?

Integration testing suggests the new negotiator does indeed work insofar as the manager doesn't spew errors about unsupported Ingress versions when the old versions are enabled (which it did if you enabled those along with the old non-negotiated controllers), but coverage parity with the old unit tests in this area would be good if possible.

Unrelated changes in the same commit include:
- A corrected KongClusterPlugin version (it was just wrong, and integration tests failed because the manager would disable the KCP controller after not seeing a v1beta1 version of that CRD).
- A change to the Knative test, loading Knative CRDs along with ours before starting the controller in suite_test.go. We need to load the Knative CRDs then so that manager doesn't auto-disable the Knative controller and break tests, though it does break with what we had before. This somewhat awkwardly loads the Knative CRDs after the cluster starts up, but doesn't load the Knative core until the Knative tests (which is where it was originally, and remains). We could move the Knative core install earlier also--it doesn't fit into the `crds` name we have for that variable, but we could reasonably rename that to something like `preloads`. Any opinions on what we should do there?

**Pull request merged**